File: cpp/markov.cpp
// markov.cpp from 'The Practice of Programming' by Pike and Kernighan

#include <iostream>
#include <fstream>
#include <deque>
#include <map>
#include <vector>
#include <cstdlib>

using namespace std;

typedef deque<string> Prefix;
map<Prefix, vector<string> > statetab; // prefix -> suffixes
enum {
    NPREF   = 2,    /* number of prefix words */
    NHASH   = 4093, /* size of state hash table array */
    MAXGEN  = 10000 /* maximum words generated */
};

const string NONWORD = "\n";

void build(Prefix& prefix, istream& in);
void add(Prefix& prefix, const string& s);
void generate(int nwords);

// markov main: markov-chain random text generation
int main(void)
{
  int nwords = MAXGEN;
  Prefix prefix;	  // current input prefix

  for (int i = 0; i < NPREF; i++) // set up initial prefix
  {
      add(prefix, NONWORD);
  }

  build(prefix, cin);
  add(prefix, NONWORD);
  generate(nwords);
  return 0;
}

// build: read input worlds, build state table
void build(Prefix& prefix, istream& in)
{
  string buf;

  while (in >> buf)
  {
      add(prefix, buf);
  }
}

// add: add word to suffix list, update prefix
void add(Prefix& prefix, const string& s)
{
  if (prefix.size() == NPREF)
  {
      statetab[prefix].push_back(s);
      prefix.pop_front();
  }
  prefix.push_back(s);
}

// generate: produce output, one word per line
void generate(int nwords)
{
  Prefix prefix;
  int i;

  for (i = 0; i < NPREF; i++) // reset initial prefix
  {
      add(prefix, NONWORD);
  }
  for (i = 0; i < nwords; i++)
  {
      vector<string>& suf = statetab[prefix];
      const string& w = suf[rand() % suf.size()];
      if (w == NONWORD)
      {
	  break;
      }
      cout << w << "\n";
      prefix.pop_front(); // advance
      prefix.push_back(w);
  }
}

File: go/markov.go
// markov.go: implementation of Markov chain text generation in Go
// Based on ideas in 'The Practice of Programming' by Pike and Kernighan

package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strings"
	"math/rand"
)

const (
	npref 		= 2 	// The number of prefixes
	maxgen 		= 10000 // The maximum generated words
	nonword		= "\n" 	// Special terminating character
)

type Prefix []string
func (p *Prefix) add(newpref string) {
	*p = append(*p, newpref)  // add newest prefix word
	*p = (*p)[1:] 		// remove oldest prefx
}

type Suffix []string

var statemap = make(map[string]Suffix)

func main() {
	build(os.Stdin)
	generate()
}

// build: read input, build prefix map
func build(file *os.File) {
	// Initialize prefix
	var pref Prefix
	for i := 0; i < npref; i++ {
		pref = append(pref, nonword)
	} 

	// Open file for reading and tokenize
	contents, err := ioutil.ReadAll(file)
	if err != nil { panic(err) }
	s := strings.Fields(string(contents))

	// Loop over input
	for i := range s {
		add(pref,s[i])
		pref.add(s[i])		
	}
	add(pref,nonword) // add end of text guard
}

// generate: produce output one word per line
func generate() {
	var pref Prefix
	for i := 0; i < npref; i++ {
		pref = append(pref, nonword)
	}
	for i := 0; i < maxgen; i++ {
		suflist := statemap[strings.Join(pref,nonword)]
		suf := suflist[rand.Intn(len(suflist))]
		if suf == nonword {
			break
		}
		fmt.Printf("%s\n",suf)
		pref.add(suf)
	}	
}

// add: add word to suffix list, update prefix
func add(pref Prefix, word string) {
	var suf []string
	key := strings.Join(pref,nonword)
	if suffix, ok := statemap[key]; ok {
		suf = suffix
	}
	statemap[key]=append(suf,word)
}
